File: Java/leetcode75/src/main/java/org/phoenix/leetcode/challenges/Problem37_LongestZigZagPathInABinaryTree.java
package org.phoenix.leetcode.challenges;

import org.phoenix.leetcode.challenges.Problem33_MaximumDepthOfBinaryTree.TreeNode;

public class Problem37_LongestZigZagPathInABinaryTree {
    int path = 0;

    public int longestZigZag(TreeNode root) {
        maxZigZag(root, "left", 0);
        maxZigZag(root, "right", 0);
        return path;
    }

    public void maxZigZag(TreeNode root, String direction, int steps) {
        if (root == null) return;
        path = Math.max(path, steps);
        if (direction == "left") {
            maxZigZag(root.left, "right", steps + 1);
            maxZigZag(root.right, "left", 1);
        } else {
            maxZigZag(root.right, "left", steps + 1);
            maxZigZag(root.left, "right", 1);
        }
    }
}

File: JavaScript/leetcode75/src/main/Problem37_LongestZigZagPathInABinaryTree.js
/**
 * @param {TreeNode} root
 * @return {number}
 */
let totalLongestPath = 0;
var longestZigZag = function (root) {
  if (root == null) return 0;
  helperZigZag(root.left, "left", 0);
  helperZigZag(root.right, "right", 0);
  return totalLongestPath;
};

let helperZigZag = function (root, direction, steps) {
  if (root == null) return;
  totalLongestPath = Math.max(totalLongestPath, steps);
  if (direction === "left") {
    helperZigZag(root.left, "right", steps + 1);
    helperZigZag(root.right, "left", 1);
  } else {
    helperZigZag(root.right, "left", steps + 1);
    helperZigZag(root.left, "right", 1);
  }
};

module.exports = longestZigZag;
